File: dotnet/GHIElectronics.DUELink/Graphics.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Drawing;
using System.Drawing.Imaging;
using System.Linq;
using System.Runtime.ConstrainedExecution;
using System.Text;
using System.Threading.Tasks;
using static GHIElectronics.DUELink.DUELinkController;

namespace GHIElectronics.DUELink {

    public partial class DUELinkController {
        public class GraphicsController {
            SerialInterface serialPort;
            StreamController stream;

            public GraphicsController(SerialInterface serialPort, StreamController stream) {
                this.serialPort = serialPort; ;
                this.stream = stream;
            }

            public bool Configuration(int type, float[] config, int width, int height, int mode) {

                //var cmd_dim_array = string.Format("dim b9[{0}]", config.Length);

                //this.serialPort.WriteCommand(cmd_dim_array);

                //var res = this.serialPort.ReadRespone();

                //for (var i = 0; i < config.Length; i++) {
                //    cmd_dim_array = string.Format("b9[{0}] = {1}", (i), config[i]);

                //    this.serialPort.WriteCommand(cmd_dim_array);

                //    res = this.serialPort.ReadRespone();

                //    if (!res.success) {
                //        break;
                //    }
                //}

                //cmd_dim_array = string.Format("dim b9[0]");

                //this.serialPort.WriteCommand(cmd_dim_array);

                //res = this.serialPort.ReadRespone();

                var config_array = string.Empty;

                config_array = "{";

                for (var i = 0; i < config.Length; i++) {
                    config_array += config[i];

                    if (i < config.Length - 1)
                        config_array += ",";
                }

                config_array += "}";

                var cmd = $"gfxcfg({type.ToString()}, {config_array}, {width}, {height}, {mode})";

                this.serialPort.WriteCommand(cmd);

                var ret = this.serialPort.ReadResponse();

                return ret.success;

            }

            public bool Show() {
                var cmd = string.Format("show()");

                this.serialPort.WriteCommand(cmd);

                var res = this.serialPort.ReadResponse();

                return res.success;
            }

            public bool Clear(uint color) {
                var cmd = string.Format("clear({0})", color);

                this.serialPort.WriteCommand(cmd);

                var res = this.serialPort.ReadResponse();

                return res.success;

            }

            

            

            public bool Pixel(uint color, int x, int y) {

                var cmd = string.Format("pixel({0},{1},{2})", color, x, y);

                this.serialPort.WriteCommand(cmd);

                var res = this.serialPort.ReadResponse();

                return res.success;
            }


            public bool Circle(uint color, int x, int y, int radius) {

                var cmd = string.Format("circle({0},{1},{2},{3})", color, x, y, radius);

                this.serialPort.WriteCommand(cmd);

                var res = this.serialPort.ReadResponse();

                return res.success;
            }

            public bool Rect(uint color, int x, int y, int width, int height) {

                var cmd = string.Format("rect({0},{1},{2},{3},{4})", color, x, y, width, height);

                this.serialPort.WriteCommand(cmd);

                var res = this.serialPort.ReadResponse();

                return res.success;
            }

            public bool Fill(uint color, int x, int y, int width, int height) {
                var cmd = string.Format("fill({0},{1},{2},{3},{4})", color, x, y, width, height);

                this.serialPort.WriteCommand(cmd);

                var res = this.serialPort.ReadResponse();

                return res.success;
            }

            public bool Line(uint color, int x1, int y1, int x2, int y2) {

                var cmd = string.Format("line({0},{1},{2},{3},{4})", color, x1, y1, x2, y2);

                this.serialPort.WriteCommand(cmd);

                var res = this.serialPort.ReadResponse();

                return res.success;
            }

            public bool Text(string text, uint color, int x, int y) {
                var cmd = string.Format("text(\"{0}\",{1},{2},{3})", text, color, x, y);


                this.serialPort.WriteCommand(cmd);

                var res = this.serialPort.ReadResponse();

                return res.success;

            }
            public bool TextS(string text, uint color, int x, int y, int scalewidth, int scaleheight) {
                var cmd = string.Format("texts(\"{0}\",{1},{2},{3},{4},{5})", text, color, x, y, scalewidth, scaleheight);


                this.serialPort.WriteCommand(cmd);

                var res = this.serialPort.ReadResponse();

                return res.success;

            }

            public bool TextT(string text, uint color, int x, int y) {
                var cmd = string.Format("textT(\"{0}\",{1},{2},{3})", text, color, x, y);


                this.serialPort.WriteCommand(cmd);

                var res = this.serialPort.ReadResponse();

                return res.success;

            }
            public bool DrawImage(float[] img, int x, int y, int w, int h, int transform) => this.DrawImageScale(img, x, y, w, h, transform, 1, 1);
            public bool DrawImageScale(float[] img, int x, int y, int width, int height, int transform, int scaleWidth, int scaleHeight) {

                if (img == null) {
                    throw new ArgumentNullException("Data null.");
                }

                if (width == 0 || height == 0 || img == null || img.Length != (width * height)) {
                    throw new ArgumentException("Invalid argument.");
                }

                var cmd = string.Format("dim a9[{0}]", img.Length);

                this.serialPort.WriteCommand(cmd);
                this.serialPort.ReadResponse();

                var written = this.stream.WriteFloats("a9", img);
                



                //for (var i = 0; i < img.Length; i++) {
                //    cmd = string.Format("b9[{0}] = {1}", (i), img[i]);

                //    this.serialPort.WriteCommand(cmd);

                //    res = this.serialPort.ReadRespone();

                //    if (!res.success) {
                //        break;
                //    }
                //}

                //var img_array = string.Empty;

                //img_array = "{";

                //for (var i = 0; i < img.Length; i++) {
                //    img_array += img[i];

                //    if (i < img.Length - 1)
                //        img_array += ",";
                //}

                //img_array += "}";

                cmd = $"imgs(a9, {x}, {y}, {width}, {height}, {transform}, {scaleWidth}, {scaleHeight})";

                this.serialPort.WriteCommand(cmd);

                var res = this.serialPort.ReadResponse();

                return res.success;

            }



          


            

        }
    }





    
}

File: python/DUELink/Graphics.py
from enum import IntEnum


class GraphicsType(IntEnum):    
    I2C = 1
    SPI = 2
    Neo = 3
    Matrix5x5 = 4


class GraphicsController:
    def __init__(self, serialPort):
        self.serialPort = serialPort
    
    def Configuration(self, displayType, config, width, height, mode)->bool:

        if not isinstance(config, list) or not all(isinstance(x, int) and 0 <= x <= 255 for x in config):
            raise ValueError("Enter a list with one number into the config with a valid code for a display.")
    
        inputConfig = map(hex, config)
        
        inputConfigArray = ",".join(inputConfig)
        
        inputConfigArray = "{" + inputConfigArray + "}"
        
        cmd = f"gfxcfg({displayType}, {inputConfigArray}, {width}, {height}, {mode})"
        self.serialPort.WriteCommand(cmd)
        res = self.serialPort.ReadResponse()
        return res.success
        
    def Show(self)->bool:
        cmd = "show()"
        self.serialPort.WriteCommand(cmd)
        res = self.serialPort.ReadResponse()
        return res.success

    def Clear(self, color)->bool:
        cmd = f"clear({color})"
        self.serialPort.WriteCommand(cmd)
        res = self.serialPort.ReadResponse()
        return res.success

    def Pixel(self, color, x, y)->bool:
        cmd = f"pixel({color},{x},{y})"
        self.serialPort.WriteCommand(cmd)
        res = self.serialPort.ReadResponse()
        return res.success

    def Circle(self, color, x, y, radius)->bool:
        cmd = f"circle({color},{x},{y},{radius})"
        self.serialPort.WriteCommand(cmd)
        res = self.serialPort.ReadResponse()
        return res.success

    def Rect(self, color, x, y, width, height)->bool:
        cmd = f"rect({color},{x},{y},{width},{height})"
        self.serialPort.WriteCommand(cmd)
        res = self.serialPort.ReadResponse()
        return res.success

    def Fill(self, color, x, y, width, height)->bool:
        cmd = f"fill({color},{x},{y},{width},{height})"
        self.serialPort.WriteCommand(cmd)
        res = self.serialPort.ReadResponse()
        return res.success

    def Line(self, color, x1, y1, x2, y2)->bool:
        cmd = f"line({color},{x1},{y1},{x2},{y2})"
        self.serialPort.WriteCommand(cmd)
        res = self.serialPort.ReadResponse()
        return res.success

    def Text(self, text, color, x, y)->bool:
        cmd = f"text(\"{text}\",{color},{x},{y})"
        self.serialPort.WriteCommand(cmd)
        res = self.serialPort.ReadResponse()
        return res.success
    
    def TextT(self, text, color, x, y)->bool:
        cmd = f"textt(\"{text}\",{color},{x},{y})"
        self.serialPort.WriteCommand(cmd)
        res = self.serialPort.ReadResponse()
        return res.success

    def TextS(self, text, color, x, y, scalewidth, scaleheight)->bool:
        cmd = f"texts(\"{text}\",{color},{x},{y},{scalewidth},{scaleheight})"
        self.serialPort.WriteCommand(cmd)
        res = self.serialPort.ReadResponse()
        return res.success

    # def __Stream(self, data, color_depth: int):
    #     cmd = f"stream({color_depth})"
    #     self.serialPort.WriteCommand(cmd)
    #     res = self.serialPort.ReadResponse()

    #     if res.success:
    #         self.serialPort.WriteRawData(data, 0, len(data))
    #         # time.sleep(10)
    #         res = self.serialPort.ReadResponse()

    #     return res.success

    def DrawImageScale(self, img, x: int, y: int, width: int, height: int, scaleWidth: int, scaleHeight: int, transform: int) -> bool:

        if width <= 0 or height <= 0 or len(img) < width * height:
            raise Exception("Invalid arguments")

        cmd = f"dim b9[{len(img)}]"

        self.serialPort.WriteCommand(cmd)
        res = self.serialPort.ReadResponse()

        for i in range(len(img)):
            cmd = f"b9[{(i)}] = {img[i]}"
            self.serialPort.WriteCommand(cmd)
            res = self.serialPort.ReadResponse()

            if (res.success == False):
                break

        if (res.success == True):
            cmd = f"imgs(b9, {x}, {y}, {width}, {height}, {scaleWidth}, {scaleHeight}, {transform})"

            self.serialPort.WriteCommand(cmd)
            res = self.serialPort.ReadResponse()

        cmd = "dim b9[0]"

        self.serialPort.WriteCommand(cmd)
        res = self.serialPort.ReadResponse()

        return res.success

    def DrawImage(self, img, x: int, y: int, width: int, height: int, transform: int) -> bool:
        return self.DrawImageScale(img, x, y, width, height, 1, 1, transform)